File: src/service/embeddingService.ts
import { Ollama, EmbeddingsResponse } from "ollama";

class EmbeddingService {
  private ollamaEmbeddings: Ollama | null = null;
  private EMBEDDING_MODEL_PATH: string | null = null;
  private OLLAMA_EMBEDDING_MODEL: string | null = null;

  constructor() {
    this.EMBEDDING_MODEL_PATH = process.env.EMBEDDING_MODEL_PATH!;
    this.OLLAMA_EMBEDDING_MODEL = process.env.OLLAMA_EMBEDDING_MODEL!;

    if (!this.EMBEDDING_MODEL_PATH || !this.OLLAMA_EMBEDDING_MODEL) {
      throw new Error("Ollama embedding details missing in .env");
    }
    this.ollamaEmbeddings = new Ollama({
      host: this.EMBEDDING_MODEL_PATH,
    });
  }

  async createEmbeddings(input: string) {
    return (await this.ollamaEmbeddings?.embeddings({
      model: this.OLLAMA_EMBEDDING_MODEL!,
      prompt: input,
    })) as EmbeddingsResponse;
  }
}

const embeddingService = new EmbeddingService();
export default embeddingService;

// curl http://localhost:11435/api/embeddings -d '{
//     "model": "nomic-embed-text",
//     "prompt": "Llamas are members of the camelid family"
//   }' | jq

File: fastapi/service/EmbeddingService.py
import os
from typing import List
from ollama import Client
from dotenv import load_dotenv
from utils import generate_unique_numbers
from langchain_core.documents import Document
from langchain_ollama import OllamaEmbeddings

load_dotenv("../.env")


class EmbeddingService(object):
    """docstring for EmbeddingService."""

    ollamaClient: Client
    OLLAMA_EMBEDDING_MODEL: str
    OLLAMA_EMBEDDING_MODEL_PATH: str

    def __init__(self):
        super(EmbeddingService, self).__init__()

        self.OLLAMA_EMBEDDING_MODEL = os.environ["OLLAMA_EMBEDDING_MODEL"]
        self.OLLAMA_EMBEDDING_MODEL_PATH = os.environ["OLLAMA_EMBEDDING_MODEL_PATH"]
        if not self.OLLAMA_EMBEDDING_MODEL or not self.OLLAMA_EMBEDDING_MODEL_PATH:
            raise Exception("Ollama credential missing in .env")

        self.ollamaClient = Client(host=self.OLLAMA_EMBEDDING_MODEL_PATH)

    def __createEmbeddings(self, text: str):
        res = self.ollamaClient.embeddings(
            model=self.OLLAMA_EMBEDDING_MODEL, prompt=text
        )

        return res["embedding"]

    def embeddedDocument(self, documents: List[Document]):

        embeddedDocuments: list = []

        for i in range(len(documents)):
            embeddedDocuments.append(
                {
                    "id": generate_unique_numbers(11231231231, 1001231231231312) + i,
                    "vector": self.__createEmbeddings(documents[i].page_content),
                    "metadatas": documents[i].metadata,
                    "document": documents[i].page_content,
                }
            )
        return embeddedDocuments


embeddingService = EmbeddingService()
